### ISSUE TITLE
Single Line Effect Behavior Changed as of 2024.03

### ISSUE THREAD

**Issue by MCStoff** [xLightsSequencer/xLights#4409]
I was using 2024.03 and placed a single line effect on the circles sub-model of the Showstopper Spinner and I noticed it was not behaving as it had in the past.  This morning 2024.05 was released so I installed that, rendered the sequence and the erratic behavior persisted with the update.  I recorded the behavior off the screen.  I then rolled back to 2024.02, re-rendered the sequence, and the single line effect behaved as expected. I then recorded the behavior off the screen. I also noticed the screen representations of the single line effect changed between 2024.02 and 2024.03 and this may be related to the change in behavior.

**To Reproduce**
To reproduce the error, play the subject matter sequence in both versions of xLights to see the difference.  I can send you a link to access the sequence upon request.

**Expected behavior**
Here is a link to a 20 second video showing the expected behavior found in 2024.02 and the behavior as of 2024.03 and thereafter.  2024.02 plays first in the video and the erratic behavior is found in the second half of the video.
https://drive.google.com/file/d/1pl6f5BJHaVfD2NuueumHsQj5Fxhld3q9/view?usp=sharing

**Screenshots**
See above video link.

**Versions (please complete the following information):**
 - OS: Windows 10
 - xLights versions 2024.02 (behaves properly) and 2024.03, 2024.04, 2024.05 (behavior is inconsistent)

**Additional context**
I would prefer to not publicly release the sequence I was working on when I noticed the issue.  I have loaded it to the google drive and can provide you the link upon request.

**Comment by derwin12** [comment]
Can you make a small sequence with your settings/etc so we can test and debug?

**Comment by MCStoff** [comment]
Sure. I will do that now. Sent from my iPhoneOn Mar 15, 2024, at 7:59 AM, Daryl ***@***.***> wrote:﻿
Can you make a small sequence with your settings/etc so we can test and debug?

**Comment by MCStoff** [comment]
Here's an animation using the sub-model where the errata is most evident.
Test Sequence for Line Effect.zip

| 
| 
|  | 
Test Sequence for Line Effect.zip


 |

 |

 |




MichaelShowstopper Sequences
 

    On Friday, March 15, 2024 at 07:59:36 AM CDT, Daryl ***@***.***> wrote:  
 
 


Can you make a small sequence with your settings/etc so we can test and debug?

**Comment by MCStoff** [comment]
MichaelShowstopper Sequences
 

    On Friday, March 15, 2024 at 08:22:32 AM CDT, Daryl ***@***.***> wrote:  
 
 


Uploads not working via the phone..

**Comment by cybercop23** [comment]
Michael, attatchements via emaila ren't showing up in here. You need to go into the issue itself and upload the file. From that we see, it works, but we do see a change that impacted SS render on stacked render so now sure yet...

**Comment by derwin12** [comment]
> Yeah only code changes I see were the addition of drawing the effect background on the grid. I was going to suggest turning off effect backgrounds to see if it changes behavior but it looks like we are always calculating the effect background. That seems like a big waste of time if the user has the effect background option turned off. Mainly commenting for us devs to look into improving that.

From .02 to .03 -- possibly the handling of duplicate nodes in submodels. I cant duplicate it ... waiting on a sample to verify.

**Comment by MCStoff** [comment]
As requested.
[Test Sequence for Line Effect.zip](https://github.com/xLightsSequencer/xLights/files/14616225/Test.Sequence.for.Line.Effect.zip)

**Comment by cybercop23** [comment]
We see it now. Very likely related to the change above...

**Comment by derwin12** [comment]
If it helps .. reverting just this code returns the previous result.. (SubModel.cpp)
![image](https://github.com/xLightsSequencer/xLights/assets/4643499/ef8f2d37-dd74-4ca4-b7ac-4c9d6428ffaf)

**Comment by dkulp** [comment]
rbegin and rend are the reverse iterator to find the last thing in the list.

**Comment by MCStoff** [comment]
If I had a vote in the matter, and I know I don't, but as someone who creates and uses alot of sub-models, I would vote for allowing duplicates in separate rows of a sub model. It would obviate the need for groups made up of the separate rows of a sub model of a single prop, which is now used to avoid duplicate pixel drop out. That would make for a cleaner sequencing environment because all sub models of a single prop could be hidden or displayed by double-clicking the prop name.  Of course, multiples of a single prop would still require sub-model groups outside the main model, unless the duplicates' effects are mapped using a mapping file.

MichaelShowstopper Sequences
 

    On Friday, March 15, 2024 at 11:30:19 AM CDT, Gil Jones ***@***.***> wrote:  
 
 


Yeah we might need a discussion on how we really want submodels to work when there are duplicate nodes. I initially didn't think duplicates should exist but I think there is a use case for it. Like if you construct a submodel with several rows and then use a stacked buffer I would expect it to light up each row in the submodel even if there are repeated nodes. I haven't analyzed it but I know if you remove duplicates from the submodel it changes the effect. I saw it when we experimented at Ron's AZ training.

Keith was against there being duplicates and Dan made that change which has affected how submodels with duplicates get rendered. We need to agree on how it should work and not breaking old behavior might not be the standard.

**Comment by dkulp** [comment]
@MCStoff  The problem with allowing duplicate nodes (and why Keith is against them) is what to do when the same node is mapped to different colors during rendering.  That behavior was never really "defined" and thus it ended up being somewhat random as to which color it would pick.  In many cases, that color could be "black" which could completely ruin the desired effect.   The above change attempted to define it so that if a node is mapped to multiple locations but only one of the locations has a color other than black, that color is used.   If multiple locations have a color, then the last one is used.

**Comment by cybercop23** [comment]
@MCStoff even for multiple HD props, I would do SM mapping and not groups, unless you really are trying to do per-preview which most likely many aren't.

@dkulp what I honestly don't get is how allowing it in a sub-model is different from a group of indidual submodels with single lines. The group would do the same thing with the overlaps so allowing it at the sm is no different. I wouldn't try to block them out and let them be as they are now, just with the added benefit of skipping a useless group creation.

**Comment by dkulp** [comment]
This is actually working "correctly" now.   When using a "Single Line" render style, the list of nodes is used "as is".  However, a node should only appear in the list of nodes once, but with possibly multiple coordinates.  This is the bug that was fixed.   With nodes appearing in multiple places, the list of nodes is now smaller.    In any case, this list now matches what happens if you define a node multiple times in the main Model grid.   It appears once but with multiple coords. 

To get the effect that you want, you would use "Default" render style, rotate 90degrees (due to vertical check box, not needed for horizontal submodel), and click on the "Group by arches".    If you slow things down so you can actually see it, you can then see the exact problem that the changes fixes.    With the old version of xLights, a bunch of the nodes in the pattern are rendered "black".  We the new submodel layout, the pattern is fully rendered and looks a ton better.

<img width="296" alt="Old" src="https://github.com/xLightsSequencer/xLights/assets/91569/2b63eb5b-6a90-4e50-ae02-3984623aac8f">
<img width="302" alt="New" src="https://github.com/xLightsSequencer/xLights/assets/91569/2a5151ae-1af4-413e-9627-582e23312d6e">

**Comment by MCStoff** [comment]
I re-installed 2024.05 and tried Dan's configuration.  It does work, but because it is not backwards compatible, it affects more than 5,000 sequences currently purchased from Showstopper Sequences and countless other sequence vendors who use the Showstopper Spinner.  So potentially thousands of end users are impacted. In addition, it is not simply set the render style to default, group the arches and rotate it 90 degrees; an effect, rotating 3d circle, used to work when the number of chases was set to 11 now requires the number of chases be set to 13.  I don't know how many other sub-models will need to be changed to get to what was originally sequenced.

If that is the solution, I suppose I can prepare a video telling my customers how to fix what was broken, once I figure out everything that was broken.

MichaelShowstopper Sequences
 

    On Friday, March 15, 2024 at 02:38:06 PM CDT, Daniel Kulp ***@***.***> wrote:  
 
 


This is actually working "correctly" now. When using a "Single Line" render style, the list of nodes is used "as is". However, a node should only appear in the list of nodes once, but with possibly multiple coordinates. This is the bug that was fixed. With nodes appearing in multiple places, the list of nodes is now smaller. In any case, this list now matches what happens if you define a node multiple times in the main Model grid. It appears once but with multiple coords.

To get the effect that you want, you would use "Default" render style, rotate 90degrees (due to vertical check box, not needed for horizontal submodel), and click on the "Group by arches". If you slow things down so you can actually see it, you can then see the exact problem that the changes fixes. With the old version of xLights, a bunch of the nodes in the pattern are rendered "black". We the new submodel layout, the pattern is fully rendered and looks a ton better.
Old.png (view on web)New.png (view on web)

**Comment by dkulp** [comment]
We (definitely Keith) have been telling everyone for years to NOT use duplicate nodes in submodels and to not rely on any behavior if doing so.   If you did, that's on you.    It's still being warned about in the check sequence for now (waiting on a rebase of a PR).   I'm not spending time trying to migrate things for something we've been telling people not to do.  If someone wants to try and tackle it, fine, I'll happily review a PR.

**Comment by MCStoff** [comment]
Because of the number of users impacted, this would be an excellent candidate for a default "old" style option as we have seen with the twinkle and pinwheel.  The effects on the sub-model do look better, but at a heavy cost.

MichaelShowstopper Sequences
 

    On Friday, March 15, 2024 at 02:38:06 PM CDT, Daniel Kulp ***@***.***> wrote:  
 
 


This is actually working "correctly" now. When using a "Single Line" render style, the list of nodes is used "as is". However, a node should only appear in the list of nodes once, but with possibly multiple coordinates. This is the bug that was fixed. With nodes appearing in multiple places, the list of nodes is now smaller. In any case, this list now matches what happens if you define a node multiple times in the main Model grid. It appears once but with multiple coords.

To get the effect that you want, you would use "Default" render style, rotate 90degrees (due to vertical check box, not needed for horizontal submodel), and click on the "Group by arches". If you slow things down so you can actually see it, you can then see the exact problem that the changes fixes. With the old version of xLights, a bunch of the nodes in the pattern are rendered "black". We the new submodel layout, the pattern is fully rendered and looks a ton better.
Old.png (view on web)New.png (view on web)

**Comment by cybercop23** [comment]
We were able  to get it to work by placing a place holder ,, where the "duplicate" nodes are on 2+ layers. That made it work as expected. We still loose the pixels which "we" don't believe we should because they (majority) of the time won't be on at the same time for all layers. Daryl is working on a simple scale-down sequence to showcase why we don't think/want to lose the pixels. Then we should schedule a zoom room review so all "sides" can state they case and discuss.

**Comment by derwin12** [comment]
With the duplicates removed out of the buffer the submodel "arms" are unequal lengths. Can placeholders be put in here somehow programatically? (i dont see an issue with the original code btw)
Nice smooth pixels ..
![image](https://github.com/xLightsSequencer/xLights/assets/4643499/20316967-1fee-4795-a4b6-fb558e8f4699)
Weird movements .. 
![image](https://github.com/xLightsSequencer/xLights/assets/4643499/acd2da2b-7ea8-41fd-ab0e-501c8b419aa0)
We thought the original code was working - so we wanted the warning removed in the check sequence.
Simple model/demo.
[4409new.zip](https://github.com/xLightsSequencer/xLights/files/14624124/4409new.zip)

**Comment by dkulp** [comment]
I can look tomorrow (on soccer fields the rest of the day today) but an on effect should be setting everything and, even if not, it should be traversing the iterator to to find a non-black.

**Comment by jackyelliott** [comment]
For what its worth, we do expect the duplicate node to get skipped but the duplicate keeps the timing on the effects. The new way seems to remove the duplicates completely and this detracts from the timing of the submodels. I understand that its not the correct way but it wasn't broken tho. Hoping we can get this to work back where it was at. Really need that timing to keep the effects working the way they have been.

**Comment by jackyelliott** [comment]
![image](https://github.com/xLightsSequencer/xLights/assets/67920202/37a218e2-b012-43d1-877b-0e07dec299ec)

**Comment by jackyelliott** [comment]
The option for removing duplicates has been there for a while now. Should it not be up to the user to determine which way is best for their application?  This function if I'm not mistaken puts a place holder instead of removing the space.

**Comment by cybercop23** [comment]
The "issue" could be _fixed_ by using the Remove Duplicates All Left->Right (guess cuz of the vertical buffer) and then Uniform Row Length.
HOWEVER... that sill makes an assumption that we want the pixels skipped, which we don't.
Below is frame 1 of using a group with individual circles that have dups, vs sub-model with dedup and equal length.
Note how with the submodel (using the new logic) the middle is almost right vs a group, however every "spoke" is missing 3 pixels which are the dups removed, in this case they should not be missing; yet the group has 3 complete spokes (10-12 o'clock) Granted as fast as it moves, you can't easily tell, but if you look close, you'll see it looks odd.
![image](https://github.com/xLightsSequencer/xLights/assets/23619433/ca70d0d3-465d-4a81-b0e0-8ace938fdc22)

**Comment by merryoncherry** [comment]
Because it is a vertical buffer, you'd have to pivot the rows and columns, then suppress duplicates left->right, then pivot it back to a vertical buffer.

**Comment by cybercop23** [comment]
Thanks, that was my thought. I think it would make it easier to understand, if regardless of buffer type, if the actions were based on the visual grid. So if a vertical buffer is set, adjust the logic but let the user "think" it is still top->bottom.
(but i digress.... :) )

**Comment by merryoncherry** [comment]
Not a bad idea.  
I'm still trying to think if there's any way to do what the OP really wants... I understand the difficulty, but that'd be very nice.

**Comment by bravado67** [comment]
I am thankful I sequence the way I always have.  When I make submodels today, I have two per one with the rows and splits I place in a group. This is because of what Merry on Cherry made possible with node suppression. I never adopted the single-line method for submodels. I would instead put the base submodels in a group or spits in a group. The beauty of xLights is the ability to do the same thing in many different ways. I'm sure it will get sorted.

**Comment by derwin12** [comment]
Good comment - but what has happened is we eliminated one of those choices and now you must use groups. Needlessly creating groups and cluttering up your layout/model list when it was tucked away in the model definition.
Anyways .. here is the pixel buffer for the buffer vs the model .. why should it be any different? If pixels are not dropped in the group version, why drop them in the submodel version.
(attached sequence for testing/poc)
![image](https://github.com/xLightsSequencer/xLights/assets/4643499/6788164f-77ef-48de-98c8-7e05f85e8283)
![image](https://github.com/xLightsSequencer/xLights/assets/4643499/f926a63f-3c8c-47db-93f4-d725936fa269)
[4409groupVSmodel.zip](https://github.com/xLightsSequencer/xLights/files/14667172/4409groupVSmodel.zip)

**Comment by bravado67** [comment]
I never saw the clutter as a problem. Most will map and move on. A model that has submodels is still cluttered if you have many, and can only see them when you reveal them from the main model. I'm just a group kind of guy.

**Comment by cybercop23** [comment]
While there are multiple ways of doing things, there's only one most efficient/best way; groups isn't it. This also doesn't help with the large growth of HD models and 10s-100s of submodels. And that shouldn't stop just because of an arbitrary limitation.

**Comment by dkulp** [comment]
SubModels are MODELS and not groups and thus should behave like a model.   They were originally designed specifically to reduce (almost eliminate) the need for Shadow models.   They should behave exactly as if you entered the nodes into  the grid of a Custom model that is a shadow model of the original.   That's exactly how it works now.

**Comment by dkulp** [comment]
From a rendering standpoint, groups are definitely the more efficient/best way.    Leverages threads/cores much better than submodels.   

Is there a particular "effect look" that needs the old "Single Line" style?   No one has shown me one yet.  In general, using "Single Line" is a really bad Render Style.  There are a lot of places in the rendering that perform poorly on extreme wide/tall but thin render buffers.

**Comment by cybercop23** [comment]
> I'm not sure I follow the different sequencing techniques but is it that some people are creating submodels that are really a collection of submodels? Like you have an Arms submodel where every row contains an arm of the model grouped into a single submodel called arms? And other sequencers were creating groups that contained each arm?

Exactly, yes... so "old" school was to create Arm1, Arm2, Arm3.. Arm 12.. then put that into a group.
New/better way is to create a submodel called Arms and either join the individual arms or have just one that has layers.
If you created a group and it had overlaps, you could silence that, but if you progressed and stepped up the modeling with layers, you are STUCK.
![image](https://github.com/xLightsSequencer/xLights/assets/23619433/915249a9-4da6-4f54-9b97-f2cc17c4ac88)

Layered Submodel
![image](https://github.com/xLightsSequencer/xLights/assets/23619433/0d3d1343-6ff5-4818-996f-f4702c511565)
vs Groups (that would have been put into a group
![image](https://github.com/xLightsSequencer/xLights/assets/23619433/5522e032-d2c5-4f91-8300-01fe15a550a4)

**Comment by cybercop23** [comment]
May I suggest that this group gathers tonight at the usual 8PM EDST in the main Zoom room to review/discuss this? Or maybe pick a diff date/time. Regular events are starting next week, so tonight's free.

**Comment by bravado67** [comment]
I have a Zoom Room that is available if you want to avoid having others popping in. Getting together and going over the different ways we sequence could be beneficial. I'm happy to help.

**Comment by cybercop23** [comment]
Thanks. We can do breakout rooms in the main xLights zoom room. Can lock if needed, but don't see a need.

**Comment by MCStoff** [comment]
This is just for background...when we created the sub-models for the Showstopper Spinner, we recognized that in stacking the sub models into "groups", there were some that had significant pixel overlap.  In the model, we broke those sub-models out into individual sub-models so the sequencer could then group those outside the model should they choose to do so.  That's why out of the 70 or so numbered sub-models, only a handful are broken out.  With the new sub-model tools that make breaking the sub-model groups out very easy, that probably wasn't necessary but was done solely as a convenience for the users of the prop.

**Comment by dkulp** [comment]
I also want to point out the sequencing at (sub)model level or in groups is almost irrelevant for this.   This ONLY affects the "Single Line" render buffer style.   The other styles are not impacted at all.     In general, using "Single Line" on anything that has multiple "rows" is a bad idea to begin with as that then ties the effect directly to that model and mapping that effect onto a different model will produce very poor results.

As an example, the effect in the sequence above has "12" chases due to the 12 circles.  If you try to map that onto a prop/submodel that only has 8, it looks horrible and requires a lot of adjusting.    Using the "default" render buffer style and an effect that doesn't rely on the count of rows is a much better option and allows the effect to map cleanly onto a much wider range of models.   For example, using Gil's favorite Morph Effect (straight up) would achieve a similar look, but would be portable to other models with different numbers of loops, or arms, or whatever.   Even a Spirals (0 wraps) on a 90 rotated buffer would look similar but be more mappable.

**Comment by MCStoff** [comment]
> I recognize its not good to change sequence behavior for users. I probably wouldn't be happy if it affected all my sequences. I also understand Dan's point that submodels were intended to operate like models. I'm available to discuss but I think we need some ideas to toss around. Seems like the goal most here want would be a way that existing sequences could still work as they did prior to 2024.03.

Worst case scenario for me is three to four days work if things stay as they are. I would go back into my existing sequences and put an off command over the old Single Line effect and put the new settings above that off command. My customers can always re-download and just import the affected sub-models (not all are affected, just those with significant pixel duplication in the sub-model).  Of course, a short training video on how to make the change to the settings would help.  I agree that backward compatibility would be ideal, but if whatever is decided makes the program better in the long-run, I will do what's necessary to make sure my customers are happy. 

Using the default render buffer with the arches grouped and the buffer rotated 90 degrees for vertical buffers looks better.  Per Dan's comment, it will make mapping easier for people using other props.  Those are both good things.  The only issue I see is the backward compatibility. Is there a way to have your cake and eat it too?

**Comment by Darzoom** [comment]
I don't like the clutter of the Groups created for the different parts of a high def prop/model.  For me, it creates no value and complicates understanding.  I prefer using submodel joins and have been doing that long before the Showstopper props/models were introduced.  My choice would be as Gil stated "Seems like the goal most here want would be a way that existing sequences could still work as they did prior to 2024.03."  Then it doesn't change anything as sequencers use both methods.

**Comment by cybercop23** [comment]
Okay... so middle of the road.. if approved, we can add logic to the SS effect to do this automagically (we've done it in other cases).. if ver>2024.5 & SS and layered SM => groups arches, change to default and if Ver buff => rotate. Nothing will need to change in models/sm or sequence.

**Comment by derwin12** [comment]
(alex beat me to it :-)
If I understand Dan's comments, the model can stay as is .. can a code change can be implemented that checks:

```
If xlights > version 2024.05
if single strand effect and single line render style and a "multilined" submodel..
then switch it to default render style and click the "group all arches"
Additionally if the submodel was marked as Vertical buffer, then rotate 90.
```
Seems like a one-time effect change like has been done previously when effect behavior has changed.

**Comment by dkulp** [comment]
It's actually slightly easier now (on master/nightlies) as SubModels now have "Horizontal" and "Vertical" render buffer styles.   If the "Default" is already vertical, vertical does nothing and horizontal will rotate.  If the default is Horizontal, vertical will rotate and horizontal does nothing.   Thus, instead of setting to "Default", we can set to "Horizontal" and it would work for both cases.

**Comment by merryoncherry** [comment]
Good discussion last night on Zoom.
I have a draft PR to address this here: https://github.com/xLightsSequencer/xLights/pull/4420
It's not final code, but it does fix the OP without being on the code path for much else, and therefore might be a good starting point for discussion.

**Comment by cybercop23** [comment]
So Dan did the automagic map and also fixed it on import. We just need to test the crap out of this using the nightly to make sure everything else works.

![image](https://github.com/xLightsSequencer/xLights/assets/23619433/e7f1218c-b2e5-4518-9d96-b3934550b909)

**Comment by dkulp** [comment]
I does a simple mapping to make it look correct.  However, the sequences should replace the effects and set the num chases to "1", turn off the "Group Arches" (which I'm going to rename to "Group Strands")  and then adjust the chase size and cycles to compensate.    Those are changes that would be very hard to do automatically.   However, that would result in effects that can be imported onto different sized props (like dropping onto the "Feathers" of a RosaGrande which has 16 feathers instead of 12 loops).   It also allow the effect on the timeline to more properly reflect what is happening.

**Comment by merryoncherry** [comment]
I tried master and it doesn't look right for number of chases like 11 or 13.  The code I have for single line gives much better results.  I'm not sure why anyone would want the new one, really.

**Comment by merryoncherry** [comment]
Near as I can tell a lot of these suggestions were about interesting ways to make all the strands do the same chase.  Some of what OP does is stagger the chases on the submodel strands, that's where having 11 or 13 chases on 12 submodels is interesting.

**Comment by derwin12** [comment]
> If you need me to test anything, I can make some time.

You would need to have Dan put you on the test flight builds .. or perhaps he can post one on his website?
